src/main/drivers/pwm_output_dshot_shared.h

```diff
@@ -0,0 +1,230 @@
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef USE_DSHOT
+
+#if defined(STM32F4) || defined(STM32F7)
+typedef DMA_Stream_TypeDef dmaStream_t;
+#else
+typedef DMA_Channel_TypeDef dmaStream_t;
+#endif
+
+FAST_RAM_ZERO_INIT static uint8_t dmaMotorTimerCount = 0;
+#ifdef STM32F7
+FAST_RAM_ZERO_INIT static motorDmaTimer_t dmaMotorTimers[MAX_DMA_TIMERS];
+FAST_RAM_ZERO_INIT static motorDmaOutput_t dmaMotors[MAX_SUPPORTED_MOTORS];
+#else
+static motorDmaTimer_t dmaMotorTimers[MAX_DMA_TIMERS];
+static motorDmaOutput_t dmaMotors[MAX_SUPPORTED_MOTORS];
+#endif
+
+#ifdef USE_DSHOT_TELEMETRY
+uint32_t readDoneCount;
+
+// TODO remove once debugging no longer needed
+FAST_RAM_ZERO_INIT uint32_t dshotInvalidPacketCount;
+FAST_RAM_ZERO_INIT uint32_t inputBuffer[DSHOT_TELEMETRY_INPUT_LEN];
+FAST_RAM_ZERO_INIT uint32_t setDirectionMicros;
+#endif
+
+motorDmaOutput_t *getMotorDmaOutput(uint8_t index)
+{
+    return &dmaMotors[index];
+}
+
+uint8_t getTimerIndex(TIM_TypeDef *timer)
+{
+    for (int i = 0; i < dmaMotorTimerCount; i++) {
+        if (dmaMotorTimers[i].timer == timer) {
+            return i;
+        }
+    }
+    dmaMotorTimers[dmaMotorTimerCount++].timer = timer;
+    return dmaMotorTimerCount - 1;
+}
+
+
+FAST_CODE void pwmWriteDshotInt(uint8_t index, uint16_t value)
+{
+    motorDmaOutput_t *const motor = &dmaMotors[index];
+
+    if (!motor->configured) {
+        return;
+    }
+
+    /*If there is a command ready to go overwrite the value and send that instead*/
+    if (pwmDshotCommandIsProcessing()) {
+        value = pwmGetDshotCommand(index);
+#ifdef USE_DSHOT_TELEMETRY
+        // reset telemetry debug statistics every time telemetry is enabled
+        if (value == DSHOT_CMD_SIGNAL_LINE_CONTINUOUS_ERPM_TELEMETRY) {
+            dshotInvalidPacketCount = 0;
+            readDoneCount = 0;
+        }
+#endif
+        if (value) {
+            motor->requestTelemetry = true;
+        }
+    }
+
+    motor->value = value;
+
+    uint16_t packet = prepareDshotPacket(motor);
+    uint8_t bufferSize;
+
+#ifdef USE_DSHOT_DMAR
+    if (useBurstDshot) {
+        bufferSize = loadDmaBuffer(&motor->timer->dmaBurstBuffer[timerLookupChannelIndex(motor->timerHardware->channel)], 4, packet);
+        motor->timer->dmaBurstLength = bufferSize * 4;
+    } else
+#endif
+    {
+        bufferSize = loadDmaBuffer(motor->dmaBuffer, 1, packet);
+        motor->timer->timerDmaSources |= motor->timerDmaSource;
+#ifdef STM32F7
+        LL_EX_DMA_SetDataLength(motor->dmaRef, bufferSize);
+        LL_EX_DMA_EnableStream(motor->dmaRef);
+#else
+        DMA_SetCurrDataCounter(motor->dmaRef, bufferSize);
+        DMA_Cmd(motor->dmaRef, ENABLE);
+#endif
+    }
+}
+
+
+#ifdef USE_DSHOT_TELEMETRY
+
+static void enableChannels(uint8_t motorCount);
```

Comment:
Just move the function up here to avoid having to use a forward declaration.

Reply:
The problem is that the function is called from the shared part of the implementation. If we move to two compilation units we'll also have this problem but we won't be able to use a static function.